tools/haskell_module_worker/src/Compile.hs

```diff
@@ -0,0 +1,279 @@
+        where phase = startPhase suff
+              (slurp,rest) = break (== "-x") args
+              these_srcs = zip slurp (repeat (Just phase))
+partition_args (arg:args) srcs objs
+  | looks_like_an_input arg = partition_args args ((arg,Nothing):srcs) objs
+  | otherwise               = partition_args args srcs (arg:objs)
+
+looks_like_an_input :: String -> Bool
+looks_like_an_input m =  isSourceFilename m
+                      || looksLikeModuleName m
+                      || "-" `isPrefixOf` m
+                      || not (hasExtension m)
+
+-- Some of this function is copied from @compiler/main/DynFlags.hs@.
+-- https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-8.10.7-release/compiler/main/DynFlags.hs#L2216
+renderLog
+  :: (String -> IO a)
+  -> DynFlags
+  -> WarnReason
+  -> Severity
+  -> SrcSpan
+  -> PprStyle
+  -> MsgDoc
+  -> IO a
+renderLog k dflags reason severity srcSpan pprStyle msg
+    = (>>= k) $ case severity of
+      SevWarning     -> printWarns
+      SevError       -> printWarns
+      _      -> return $ showSDoc dflags msg
+    where
+      message = mkLocMessageAnn flagMsg severity srcSpan msg
+
+      printWarns = do
+        caretDiagnostic <-
+            if gopt Opt_DiagnosticsShowCaret dflags
+            then getCaretDiagnostic severity srcSpan
+            else pure empty
+        return $ showSDoc dflags $
+          withPprStyle pprStyle (message $+$ caretDiagnostic)
+
+      flagMsg =
+        case reason of
+          NoReason -> Nothing
+          Reason wflag -> do
+            spec <- flagSpecOf wflag
+            return ("-W" ++ flagSpecName spec ++ warnFlagGrp wflag)
+          ErrReason Nothing ->
+            return "-Werror"
+          ErrReason (Just wflag) -> do
+            spec <- flagSpecOf wflag
+            return $
+              "-W" ++ flagSpecName spec ++ warnFlagGrp wflag ++
+              ", -Werror=" ++ flagSpecName spec
+
+      warnFlagGrp flag
+          | gopt Opt_ShowWarnGroups dflags =
+                case smallestGroups flag of
+                    [] -> ""
+                    groups -> " (in " ++ intercalate ", " (map ("-W"++) groups) ++ ")"
+          | otherwise = ""
+
+-- | Find the 'FlagSpec' for a 'WarningFlag'.
+flagSpecOf :: WarningFlag -> Maybe (FlagSpec WarningFlag)
+flagSpecOf flag = listToMaybe $ filter check wWarningFlags
+  where
+    check fs = flagSpecFlag fs == flag
+
+-- | Find the smallest group in every hierarchy which a warning
+-- belongs to, excluding Weverything.
+smallestGroups :: WarningFlag -> [String]
+smallestGroups flag = mapMaybe go warningHierarchies where
+    -- Because each hierarchy is arranged from smallest to largest,
+    -- the first group we find in a hierarchy which contains the flag
+    -- is the smallest.
+    go (group:rest) = fromMaybe (go rest) $ do
+        flags <- lookup group warningGroups
+        guard (flag `elem` flags)
+        pure (Just group)
+    go [] = Nothing
```

Comment:
It seems to me like you return a `Maybe (Maybe a)` here, but the value can never be `Just Nothing`. So this can be simplified to
```haskell
smallestGroups :: WarningFlag -> [String]
smallestGroups flag = mapMaybe go warningHierarchies where
    -- Because each hierarchy is arranged from smallest to largest,
    -- the first group we find in a hierarchy which contains the flag
    -- is the smallest.
    go :: [String] -> Maybe String
    go (group:rest) = do
        flags <- lookup group warningGroups
        if flag `elem` flags
        then Just group
        else go rest
    go [] = Nothing
```
Additionally, I also suspect you can slightly simplify this further by using `find` instead:
```haskell
-- Because each hierarchy is arranged from smallest to largest,
-- the first group we find in a hierarchy which contains the flag
-- is the smallest.
smallestGroups :: WarningFlag -> [String]
smallestGroups flag = mapMaybe (find go) warningHierarchies where
    go :: String -> Bool
    go group =
      case lookup group warningGroups of
        Nothing -> False
        Just flags -> flag `elem` flags
```

Reply:
Thanks @googleson78. This code is copied from GHC though, so I rather leave it as is.

Reply:
Ah, of course :sweat_smile: